project13/ViewController.swift

```diff
@@ -4,16 +4,191 @@
 //
 //  Created by Felipe Gil on 2021-07-09.
 //
-
+import CoreImage
 import UIKit
+//TODO
+#warning("Make the Change Filter button change its title to show the name of the currently filter")
+#warning("Try more than 1 slider with different properties")
 
-class ViewController: UIViewController {
-
+class ViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
+    @IBOutlet weak var imageView: UIImageView!
+    @IBOutlet weak var intensity: UISlider!
+    
+    var currentImage: UIImage?
+    lazy var context: CIContext = CIContext()
+    var currentFilter: CIFilter!
+    
+    let loadingActivityIndicator: UIActivityIndicatorView = {
+        let indicator = UIActivityIndicatorView()
+        indicator.style = .large
+        indicator.color = .white
+        indicator.startAnimating()
+        indicator.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin, .flexibleTopMargin, .flexibleBottomMargin]
+        return indicator
+    }()
+    let blurEffectView: UIVisualEffectView = {
+        let blurEffect = UIBlurEffect(style: .dark)
+        let blurEffectView = UIVisualEffectView(effect: blurEffect)
+        blurEffectView.alpha = 0.9
+        blurEffectView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+        return blurEffectView
+    }()
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        
+        let newTitle: String = "InstaFilter"
+        title = newTitle
+        
+        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(importPicture))
+
+        currentFilter = CIFilter(name: "CITwirlDistortion")
+        
+    }
+    @objc private func importPicture() {
+        startLoadingAnimation()
+        DispatchQueue.main.async { [weak self] in
+            guard let self = self else { return }
+            let picker = UIImagePickerController()
+            picker.allowsEditing = true
+            picker.delegate = self
+            self.present(picker, animated: true) {
+                self.stopLoadingAnimation()
+            }
+        }
+    }
+    
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        
+        guard let image = info[.editedImage] as? UIImage else { return }
+        dismiss(animated: true)
+        currentImage = image
+        guard let currentImage = currentImage else { return }
+        
+        let beginImage = CIImage(image: currentImage)
+        currentFilter.setValue(beginImage, forKey: kCIInputImageKey)
```

Comment:
The code in line 66 doesn't make much sense since you set the value of `currentImage` in the immediate line above. In cases like this we usually use the local var (`image`) where it's needed (line 68) and update the global one only in the end (`currentImage`).

```suggestion
        currentFilter.setValue(CIImage(image: image), forKey: kCIInputImageKey)
        currentImage = image
```

Reply:
Fixed!